File: src/dnd/effects/Dodge.java
package dnd.effects;

import dnd.events.Event;
import dnd.events.dicecontests.attackroll.AttackRoll;
import dnd.events.dicecontests.savingthrow.SavingThrow;
import gameobjects.entities.Entity;

public class Dodge extends Effect {

	public Dodge(Entity source, Entity target) {
		super(source, target, "Dodge", HI_PRIORITY);
	}

	@Override
	public boolean processEvent(Event e, Entity source, Entity target) {
		if (e instanceof AttackRoll && target == getTarget()) {
			try {
				e.applyEffect(this);
				((AttackRoll)e).grantDisadvantage(this);
				return true;
			} catch (Exception ex) {}
		}
		else if (e instanceof SavingThrow && target == getTarget() && ((SavingThrow)e).getSaveAbility() == Entity.DEX) {
			try {
				e.applyEffect(this);
				((SavingThrow)e).grantAdvantage(this);
				return true;
			} catch (Exception ex) {}
		}
		return false;
	}

}

File: scripts/effects/dodge.lua
function define()
	self:setName("Dodge")
end



function processEvent()
	Entity = luajava.bindClass("com.dndsuite.core.gameobjects.Entity")
	DiceContest = luajava.bindClass("com.dndsuite.core.events.contests.DiceContest")
	AttackRoll = luajava.bindClass("com.dndsuite.core.events.contests.AttackRoll")
	SavingThrow = luajava.bindClass("com.dndsuite.core.events.contests.SavingThrow")
	
	if target == self:getTarget() then
		if event:hasTag(AttackRoll:getEventID()) then
			event:applyEffect(self)
			event:addTag(DiceContest.DISADVANTAGE)
		elseif event:hasTag(SavingThrow:getEventID()) and event:getSaveAbility() == Entity.DEX then
			event:applyEffect(self)
			event:addTag(DiceContest.ADVANTAGE)
		else
			print("[LUA]  No changes made to " .. event:toString())
		end
	else
		print("[LUA]  No changes made to " .. event:toString())
	end
end



function processEventSafe()
	print("[LUA]  Effect " .. self:toString() .. " processing event " .. event:toString() .. "...")
	if not pcall(processEvent) then
		print("[LUA]  Effect " .. self:toString() .. " already applied to event " .. event:toString())
	end
end